### PR TITLE
DE improvement

### PR DESCRIPTION
**Pull request by JoeWaddell**
## Background ##
The NASEM_Comp.csv doesn't include DE values, but DE_Base. In essence, DE is not supplied directly for NASEM (it is in the NRC_Comp.csv). Instead, the NASEM standard intends for users to calculate DE using DE_Base and other values in an animal's ration - however implementation in RuFaS is a later milestone, see Issue #834.

## Reason for this short PR ##
There is a slight discrepancy in many DE v. DE_Base values, leading to incorrect rejection of user-defined rations. I noticed this testing another branch, and found that it will fail energy constraints, which were using DE_Base values. 

## fix ##
Because calculation of DE from DE_Base is pending changes in Issue #834, for now I've implemented a quick change, similar to the fix for TDN previously implemented in PR #892, using the DE values from the NRC database where found. For those not found in the NRC database I entered '-1', and added code to use DE only if the value is not '-1'.

### CODE REVIEW

File: input/data/feed/NASEM_Comp_with_TDN.csv [hunk 1]
```diff
@@ -1,165 +1,165 @@
...
+203,Milk replacer 28 CP 17.5 fat,Calf Liquid Feed,Conc,15.0,7.0,29.2,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,45.6,18.0,18.2,4.93,0.8,0.6,0.15,1.1,0.4,0.32,0.2,5.0,85.0,60.0,65.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,80,0,0,0,0.1,1000,10,9800,3200,200,0.0,0.0,0,0.0,0,0.0,0.0,0.0,0.0,0.0,0,0,kg,100,0,0.0,4.78
+201,"Skim milk, powder",Calf Liquid Feed,Conc,94.0,6.9,37.4,0,0.0,0.0,0.0,0.0,100.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,56.0,0.8,1.0,4.25,1.29,1.08,0.1,1.85,0.55,1.0,0.45,0.42,3.3,0.2,42.0,0.0,0.0,0.01,0.15,0.2,3.75,2.92,6.18,10.6,8.8,3.0,5.26,4.62,1.65,6.9,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0,80,0,100,0,0.0,0,0,0,0,0,0.0,0.0,0,0.0,0,0.0,0.0,0.0,0.0,0.0,0,0,kg,100,0,0.0,0.0
```
Comment (Ed731551) [inline input/data/feed/NASEM_Comp_with_TDN.csv]: Please replace DE value from 0 to 4.25 in feed ID 201

File: input/data/feed/NASEM_Comp_with_TDN.csv [hunk 2]
```diff
@@ -1,165 +1,165 @@
...
+298,Vit D premix,Vitamin/Mineral,Conc,100.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,91,73,100,0,0,0.0,0,0,0,250000,0,0.6,0.68,1,0.92,1,0.05,0.1,0.3,0.004,0.2,0,0,kg,100,0,0.0,-1.0
+299,Vit E premix,Vitamin/Mineral,Conc,100.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,91,73,100,0,0,0.0,0,0,0,0,44000,0.6,0.68,1,0.92,1,0.05,0.1,0.3,0.004,0.2,0,0,kg,100,0,0.0,-1.0
+300,"VitTM Premix, generic",Vitamin/Mineral,Conc,92.0,18.0,16.0,40,50.0,10.0,3.38,43.054,85.033,18.222,1.0,3.5,12.0,26.0,0.0,4.5,15.0,10.0,2.0,3.0,2.597,0.8,0.8,0.4,1.0,0.021,0.04,0.6,1600.0,1000.0,4000.0,4000.0,0.0,0.0,0.0,80.0,50.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,91,73,100,0,0,0.0,0,0,560000,200000,4000,0.5,0.68,1,0.92,1,0.05,0.1,0.23,0.005,0.2,0,0,kg,100,0,0.0,-1.0
```
Comment (Ed731551) [inline input/data/feed/NASEM_Comp_with_TDN.csv]: DE base in Feed ID 30 'VitTM Premix, generic' should be equal to zero.

**Comment by Ed731551** [comment]
Hi Joe, after revising the file NASEM_Comp_with_TDN.csv I can say the following: 

1. The DE values for all mineral feeds within the Fd_Category:  **Vitamin/Mineral** must be zero (0). That also includes the feed 300 (RuFaS ID) -' VitTM Premix, generic', which at present includes a value of  DE_Base equal to 2.597 (the last does not make sense to me, plese replace that by 0).
2. For feed, '201 Skim milk, powder', the DE value according to NRC (2001) book is equal to 4.25 (Mcal/kg DM - units) and not zero.
3. I noted that feed composition for 'Skim milk fresh' (RuFaS ID number 207) is missing. That is included defined in the file RuFaS_Feed_Library_Combined_Final.xlsx, in the BaseCamp folder:
https://3.basecamp.com/3486446/buckets/5296287/google_documents/6374467960 and not included in the csv file. Could you please revise this? That feed should be included. Anyhow its DE value according to NRC is equal to 4.19. 

Thanks.

**Comment by JoeWaddell** [comment]
Thanks @Ed731551 , replies below:

 > 1.   The DE values for all mineral feeds within the Fd_Category: Vitamin/Mineral must be zero (0). That also includes the feed 300 (RuFaS ID) -' VitTM Premix, generic', which at present includes a value of DE_Base equal to 2.597 (the last does not make sense to me, plese replace that by 0).

Updated in latest commit: but note, this is an error in the RuFaS_Feed_Library_Combined_Final.xlsx file, see the NASEM tab. 

 > 2.  For feed, '201 Skim milk, powder', the DE value according to NRC (2001) book is equal to 4.25 (Mcal/kg DM - units) and not zero.

Updated in latest commit for both NRC and NASEM Comp CSVs: but note, this is an error in the RuFaS_Feed_Library_Combined_Final.xlsx file, see the NRC tab. 
  
> 3. I noted that feed composition for 'Skim milk fresh' (RuFaS ID number 207) is missing. That is included defined in the file RuFaS_Feed_Library_Combined_Final.xlsx, in the BaseCamp folder:
    https://3.basecamp.com/3486446/buckets/5296287/google_documents/6374467960 and not included in the csv file. Could you please revise this? That feed should be included. Anyhow its DE value according to NRC is equal to 4.19.

This was not included because it is missing from the NRC Comp csv, and so we don't have a TDN to use (see issue #834). 

Note: I have not edited `RuFaS_Feed_Library_Combined_Final.xlsx`

**Comment by Ed731551** [comment]
Thanks @JoeWaddell.

**Review by Ed731551 — APPROVED** [review]
to me implemented changes are Ok. Thanks @JoeWaddell

**Review by allisterakun — APPROVED** [review]
LGTM! Please update the PR description of using '-1' instead of 'NA' for missing values!

**Pull request merged**